flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -49,6 +49,12 @@
 
 	private volatile boolean externallyInducedCheckpoints;
 
+	/**
+	 * Indicates whether this Task was purposefully finished (by finishTask()), in this case we
+	 * want to ignore exceptions thrown after finishing, to ensure shutdown works smoothly.
+	 */
+	private volatile boolean isFinished = false;
+
```

Comment:
I would suggest to move this flag to the `StreamTask`. The reason is that although for now it is used only in the `SourceStreamTask`, know if the job was normally terminated or cancelled is a property relevant to all the `StreamTask`s.